Fixing `ToolStrip` button background rendering in HC theme

ArtemTatarinov opened the pull request:
Fixes #5502

## Proposed changes

- `false` flag is being passed while creating `ToolStripHighContrastRenderer` instance at `ToolStripSystemRenderer`, so after this flag is inverted at `ToolStripHighContrastRenderer`'s constructor, the `FillWhenSelected` flag will be `true` (`ToolStripProfessionalRenderer` already has the same logic) 
- If there's a special overridden renderer for High Contrast theme, then base class should call that renderer's method for rendering backgrounds of `ToolStripButton`, `ToolStripDropDownButton`, `ToolStripSplitButton` (`ToolStripProfessionalRenderer` already has the same logic)  

## Customer Impact

**Before fix**

![image](https://user-images.githubusercontent.com/87859299/131334951-ae4081c5-fa28-44bf-a52f-476c562dd44d.png)

**After fix**
![image](https://user-images.githubusercontent.com/87859299/131335345-6306989e-a65b-4b59-b35b-9c695c19714b.png)
![image](https://user-images.githubusercontent.com/87859299/131335358-3f958c63-37c1-4f5d-8642-c6432a8d0f25.png)
![image](https://user-images.githubusercontent.com/87859299/131335363-f20dc935-bc8d-4f1c-b6fb-3c07e7850f7d.png)

## Regression? 

- No

## Risk

- Minimal

## Test methodology <!-- How did you ensure quality? -->

- Manually
- Unit tests
- CTI team

## Accessibility testing  <!-- Remove this section if PR does not change UI -->

- Inspect
- Accessibility Insights

## Test environment(s) <!-- Remove any that don't apply -->
Microsoft Windows [Version 10.0.19043.1165]
.NET SDK 6.0.100-rc.1.21416.15

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/5614)

dreddy-work commented on src/System.Windows.Forms/src/System/Windows/Forms/ToolStripSystemRenderer.cs:
```diff
@@ -363,6 +373,14 @@ protected override void OnRenderButtonBackground(ToolStripItemRenderEventArgs e)
         {
+            //if there's a special overridden renderer for High Contrast theme, then base class should call that renderer's method for rendering the button background
+            //ToolStripProfessionalRenderer has the same logic
+            if (RendererOverride != null)
```
nit: is not

dreddy-work commented on src/System.Windows.Forms/src/System/Windows/Forms/ToolStripSystemRenderer.cs:
```diff
@@ -355,6 +357,14 @@ protected override void OnRenderImageMargin(ToolStripRenderEventArgs e)
+            //if there's a special overridden renderer for High Contrast theme, then base class should call that renderer's method for rendering the button background
+            //ToolStripProfessionalRenderer has the same logic
+            if (RendererOverride is not null)
+            {
```
why can't we simplify this and call RendererOverride.OnRenderButtonBackground(e) directly?

ArtemTatarinov replied:
We're calling the base class method, because `ToolStripProfessionalRenderer` does it, so the code will be consistent.

`ToolStripProfessionalRenderer.cs`:
![image](https://user-images.githubusercontent.com/87859299/131445472-0cf5631b-6370-4921-a373-07f0214aa1d9.png)
![image](https://user-images.githubusercontent.com/87859299/131445538-94ff4b21-4700-46c5-9110-d4af02b0858f.png)
![image](https://user-images.githubusercontent.com/87859299/131445561-a48d7998-c197-4a37-bde6-4968cafaadfd.png)

dreddy-work replied:
I saw that. But i do not see a reason to call base and check for custom renderer again. We could change `ToolStripProfessionalRenderer `too.

ArtemTatarinov replied:
Some investigation showed that we can't call the `OnRenderSplitButtonBackground` method from the `ToolStripSystemRenderer` class, because this method is declared at the base abstract class `ToolStripRenderer` as `protected`:

![image](https://user-images.githubusercontent.com/87859299/131659857-1edf15f5-13c8-45a5-9186-295f17ee22a4.png)

![image](https://user-images.githubusercontent.com/87859299/131660798-dc0d6d85-df41-4623-a6c4-ba1b4e1d89db.png)

Inside the class 'A' itself we can call protected methods of the objects with the same type 'A' too, that's why we can call `protected` method `OnRenderSplitButtonBackground` from `RendererOverride` property. 

`ToolStripRenderer.cs`:
![image](https://user-images.githubusercontent.com/87859299/131661018-0664dae6-d0a1-49fc-bcad-b435ae81aa76.png)

And I guess that's why this approach was used in `ToolStripProfessionalRenderer` in the first place. But if we want to change it, we could use something like that:

![image](https://user-images.githubusercontent.com/87859299/131662800-a44b9477-fcf3-42ac-bf18-6bc2f3aff15b.png)

Will it be appropriate?

dreddy-work replied:
In that case, i will not block this PR for this. Changing modifier on base class method from protected to internal may be the right approach here.  Can you track that with an issue for cleanup?

ArtemTatarinov replied:
> In that case, i will not block this PR for this. Changing modifier on base class method from protected to internal may be the right approach here. Can you track that with an issue for cleanup?

Thank you! Issue created: https://github.com/dotnet/winforms/issues/5672

dreddy-work commented on src/System.Windows.Forms/src/System/Windows/Forms/ToolStripSystemRenderer.cs:
```diff
@@ -355,6 +357,14 @@ protected override void OnRenderImageMargin(ToolStripRenderEventArgs e)
         /// </summary>
         protected override void OnRenderButtonBackground(ToolStripItemRenderEventArgs e)
         {
+            //if there's a special overridden renderer for High Contrast theme, then base class should call that renderer's method for rendering the button background
+            //ToolStripProfessionalRenderer has the same logic
```
```suggestion
            // If system in high contrast mode and specific renderer override is defined, use that. This is in consistence with ToolStripProfessionalRenderer.
```

dreddy-work commented on src/System.Windows.Forms/src/System/Windows/Forms/ToolStripSystemRenderer.cs:
```diff
@@ -42,7 +42,9 @@ internal ToolStripRenderer HighContrastRenderer
                 if (toolStripHighContrastRenderer is null)
                 {
-                    toolStripHighContrastRenderer = new ToolStripHighContrastRenderer(/*renderLikeSystem*/true);
+                    //'false' flag is being passed to the ToolStripHighContrastRenderer constructor as for the ToolStripProfessionalRenderer,
+                    //so DottedBorder, DottedGrip, FillWhenSelected flags will be 'true' at the ToolStripHighContrastRenderer for toolstrip system renderer too
```
nit: simplify this comment.

RussKie replied:
Each comment must be in "sentence case", i.e. start with a capital letter, and end with a full stop.
Also there must be one space between "//" and the first word.

dreddy-work commented:
As per @merriemcgaw , we would need this servicing to 6.0 RC2.

Tanya-Solyanik commented on src/System.Windows.Forms/src/System/Windows/Forms/ToolStripSystemRenderer.cs:
```diff
@@ -355,6 +356,13 @@ protected override void OnRenderImageMargin(ToolStripRenderEventArgs e)
         /// </summary>
         protected override void OnRenderButtonBackground(ToolStripItemRenderEventArgs e)
         {
+            // If system in high contrast mode and specific renderer override is defined, use that. This is in consistence with ToolStripProfessionalRenderer.
```
>This is in consistence with ToolStripProfessionalRenderer.

A better comment would be to explain that `RendererOverride` is set only in HC (as far as I remember, please verify)

Tanya-Solyanik reviewed (commented):
Looks good, but please describe testing done. Specifically it is important to test renderers for all possible styles  and states of tool strip buttons, [checked](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.toolstripbutton.checked?view=net-5.0), [CHeckState](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.toolstripbutton.checkstate?view=net-5.0), and in response to different mouse  actions(mouse over, left button down).

Consider adding a test case that enumerates these variations like [buttons.cs]( https://github.com/dotnet/winforms/blob/46ba4d049c8fd3d79f15a545321af2d99b9c262f/src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/Buttons.cs#L44) does to our test app.

RussKie commented:
Generally we should write [rendering tests](https://github.com/dotnet/winforms/blob/main/docs/testing.md#rendering-tests), if possible. There are a few example in the codebase, e.g. https://github.com/dotnet/winforms/pull/4739.

Please ping me if you have any Qs.

dreddy-work reviewed (dismissed):
LGTM.

ArtemTatarinov commented:
> Looks good, but please describe testing done. Specifically it is important to test renderers for all possible styles and states of tool strip buttons, [checked](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.toolstripbutton.checked?view=net-5.0), [CHeckState](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.toolstripbutton.checkstate?view=net-5.0), and in response to different mouse actions(mouse over, left button down).
Consider adding a test case that enumerates these variations like [buttons.cs](https://github.com/dotnet/winforms/blob/46ba4d049c8fd3d79f15a545321af2d99b9c262f/src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/Buttons.cs#L44) does to our test app.

I have added `ToolStrip` buttons to the `ToolStripTests` form and manually checked different mouse actions (mouse hover, LMB and RMB clicking), different DisplayStyles (`Image`, `Text`, `ImageAndText`) and Checked/CheckStates (`Checked`, `Unchecked`, `Indeterminate`) in Black and White HC themes. I found one issue: for `CheckState.Indeterminate` the background is rendered wrong just like the original cases of the issue but this time it doesn't matter if the button was selected:
![image](https://user-images.githubusercontent.com/87859299/132017746-407aa032-b586-4fff-bd3c-b3003efe0bff.png)
![image](https://user-images.githubusercontent.com/87859299/132017957-4b7c104b-f969-460a-98ec-c55d49256b05.png)

I fixed it and tested again manually all the cases, looks like now it's good.
![image](https://user-images.githubusercontent.com/87859299/132018180-01d47bf7-166b-4c45-8501-de2fd6d3608e.png)
![image](https://user-images.githubusercontent.com/87859299/132018101-b7e65cbd-01c9-46bd-8bb9-ed3b18108a07.png)

**Issue Description**
Background of the buttons with `CheckState.Checked` and `CheckState.Indeterminate` is rendered the same way in non-HC mode at .NET 6.0 and .NET Framework 4.7.2 , both states look like each other.

.NET 6.0
![image](https://user-images.githubusercontent.com/87859299/132023227-2fccf603-ba0a-446f-990e-a22231b0bc60.png)

.NET Framework 4.7.2
![image](https://user-images.githubusercontent.com/87859299/132023119-f5a97613-3a02-4ba3-9e5c-de9e50c7cae0.png)

This is happening because `Checked` property is used for rendering the `ToolStrip` items, not `CheckState` property, and the `Checked` property of the `ToolStripButton` class is `false` only for `CheckState.Unchecked` and is `true` for `CheckState.Checked` or `CheckState.Indeterminate`:
![image](https://user-images.githubusercontent.com/87859299/132021123-173ffdfb-9cac-4be0-bc81-01c1d9f441c0.png) 

I used the same logic for fixing this issue. At the `OnRenderButtonBackground` method of the `ToolStripHighContrastRenderer` class there were two checks:
![image](https://user-images.githubusercontent.com/87859299/132020976-9099dec9-b273-436c-9b14-2fdd1794ebe8.png)

So, if we want to render the background of the button with `CheckState.Indeterminate` state in HC mode as it's done in non-HC mode, we should remove the redundant check for `CheckState.Checked`, because it will leave buttons with `CheckState.Indeterminate` without correct background rendering:
![image](https://user-images.githubusercontent.com/87859299/132025006-bcb9da19-1645-4697-acf9-58fee8a29b60.png)

Tanya-Solyanik commented:
>Background of the buttons with CheckState.Checked and CheckState.Indeterminate is rendered the same way in non-HC mode at .NET 6.0 and .NET Framework 4.7.2 , both states look like each other.

This is unfortunate... But this is an existing bug, we'll wait for external reports before we triage it. 

Your fix for indeterminate HC makes sense. Thank you for the follow up!

Tanya-Solyanik reviewed (dismissed):
Looks good

Tanya-Solyanik commented:
Issue to track rendering tests - https://github.com/dotnet/winforms/issues/5670

vladimir-krestov commented on src/System.Windows.Forms/src/System/Windows/Forms/ToolStripSystemRenderer.cs:
```diff
@@ -42,7 +42,7 @@ internal ToolStripRenderer HighContrastRenderer
                 if (toolStripHighContrastRenderer is null)
                 {
-                    toolStripHighContrastRenderer = new ToolStripHighContrastRenderer(/*renderLikeSystem*/true);
+                    toolStripHighContrastRenderer = new ToolStripHighContrastRenderer(/*renderLikeSystem*/systemRenderMode: false);
```
```suggestion
                    toolStripHighContrastRenderer = new ToolStripHighContrastRenderer(systemRenderMode: false);
```

vladimir-krestov reviewed (dismissed):
Looks good ✔️ 
Waiting for testing results...

RussKie reviewed (commented):
Rendering tests?

Olina-Zhang commented:
Tested this fixing with no new issue found, hitting known GH issue: https://github.com/dotnet/winforms/issues/5676

dreddy-work commented:
> Rendering tests?

Rendering tests are tracked by #5670 .
